### PR TITLE
Mention the section when a setting is missing

### PR DESCRIPTION
**Pull request by Nosferatul**
Mention the section name when a required setting is
missing for the user.
The user will know what section is used.

Closes https://github.com/coala/coala/issues/2589

### CODE REVIEW

**Comment by Nosferatul** [comment]
The output:
```
Section: Default
Please enter a value for the setting "use_spaces" (True if spaces are to be used instead of tabs.) needed by SpaceConsistencyBear: 
```

**Comment by Nosferatul** [comment]
Should I change the tests, right?

File: coalib/output/ConsoleInteraction.py [hunk 1]
```diff
@@ -460,7 +460,7 @@ def join_names(values):
         return values[0]
```
Comment (Adrianzatreanu) [inline coalib/output/ConsoleInteraction.py]: commit message: "return"->"returns"

File: coalib/output/ConsoleInteraction.py [hunk 2]
```diff
@@ -460,7 +460,7 @@ def join_names(values):
...
 
 
-def require_setting(setting_name, arr):
+def require_setting(setting_name, arr, section):
```
Comment (Adrianzatreanu) [inline coalib/output/ConsoleInteraction.py]: document :param section:

File: coalib/output/ConsoleInteraction.py [hunk 3]
```diff
@@ -471,14 +471,17 @@ def require_setting(setting_name, arr):
                          following.
     """
     needed = join_names(arr[1:])
+    section_name = "Section: " + section.name_str + ". "
 
     # Don't use input, it can't deal with escapes!
-    print(colored(STR_GET_VAL_FOR_SETTING.format(setting_name, arr[0], needed),
+
+    print(colored(section_name + STR_GET_VAL_FOR_SETTING.format(setting_name,
+                                                                arr[0], needed),
```
Comment (Adrianzatreanu) [inline coalib/output/ConsoleInteraction.py]: could you please post here how it looks now and how it looked before? thanks :D 2 imgurs of terminal suffice.

File: coalib/output/ConsoleInteraction.py [hunk 4]
```diff
@@ -471,14 +471,17 @@ def require_setting(setting_name, arr):
...
                   REQUIRED_SETTINGS_COLOR))
     return input()
 
 
-def acquire_settings(log_printer, settings_names_dict):
+def acquire_settings(log_printer, settings_names_dict, section):
```
Comment (Adrianzatreanu) [inline coalib/output/ConsoleInteraction.py]: same here, document the param.

File: coalib/settings/Section.py [hunk 5]
```diff
@@ -253,3 +253,7 @@ def delete_setting(self, key):
         :param key: The key of the setting to be deleted
         """
         del self.contents[key]
+
+    @property
+    def name_str(self):
```
Comment (Adrianzatreanu) [inline coalib/settings/Section.py]: i thik you need to document properties too. as parameters are missing here, you should explain what it returns, as in
:return:

**Comment by Adrianzatreanu** [comment]
yes, apparently i think you fked up somewhere when modifying the tests and they are not passing now :D

File: coalib/output/ConsoleInteraction.py (hunk 3 above)
Comment (Nosferatul) [inline coalib/output/ConsoleInteraction.py]: /home/alex/Pictures/Screenshot from 2016-10-01 04-15-02.png

Comment (Nosferatul) [inline coalib/output/ConsoleInteraction.py]: http://imgur.com/a/VgIrB

**Comment by Nosferatul** [comment]
I don't know how to change the tests :(

File: coalib/output/ConsoleInteraction.py (hunk 3 above)
Comment (sils1297) [inline coalib/output/ConsoleInteraction.py]: I'd rather put the info into the sentence somehow.

File: coalib/output/ConsoleInteraction.py [hunk 6]
```diff
@@ -489,6 +493,7 @@ def acquire_settings(log_printer, settings_names_dict):
         A dictionary with the settings name as key and a list containing a
         description in [0] and the name of the bears who need this setting in
         [1] and following.
+    :param section:         The section the action corresponds to.
```
Comment (sils1297) [inline coalib/output/ConsoleInteraction.py]: the example below this line belongs to the param described above it

File: coalib/output/ConsoleInteraction.py (hunk 3 above)
Comment (Nosferatul) [inline coalib/output/ConsoleInteraction.py]: ....`needed by <bear> from section: Default` ?

**Comment by Nosferatul** [comment]
....`needed by <bear> from section: Default` ?

File: coalib/output/ConsoleInteraction.py (hunk 3 above)
Comment (sils1297) [inline coalib/output/ConsoleInteraction.py]: I'd rather `from section "Default"` without the colon

Comment (Nosferatul) [inline coalib/output/ConsoleInteraction.py]: Sure

**Comment by Nosferatul** [comment]
Sorry for taking so long :worried:

**Comment by Nosferatul** [comment]
http://imgur.com/a/csaDA

File: coalib/output/ConsoleInteraction.py [hunk 7]
```diff
@@ -479,16 +479,20 @@ def require_setting(setting_name, arr):
     :param arr:          A list containing a description in [0] and the name
                          of the bears who need this setting in [1] and
                          following.
+    :param section:         The section the action corresponds to.
```
Comment (srisankethu) [inline coalib/output/ConsoleInteraction.py]: The params description are not in line

Comment (srisankethu) [inline coalib/output/ConsoleInteraction.py]: params description not in line

File: coalib/output/ConsoleInteraction.py [hunk 8]
```diff
@@ -508,6 +512,8 @@ def acquire_settings(log_printer, settings_names_dict):
                      "SpaceConsistencyBear",
                      "SomeOtherBear"]}
 
+    :param section:         The section the action corresponds to.
```
Comment (srisankethu) [inline coalib/output/ConsoleInteraction.py]: params description not in line

File: coalib/output/ConsoleInteraction.py [hunk 9]
```diff
@@ -508,6 +512,9 @@ def acquire_settings(log_printer, settings_names_dict):
                      "SpaceConsistencyBear",
                      "SomeOtherBear"]}
 
+    :param section:
+        The section the action corresponds to.
+
```
Comment (Adrianzatreanu) [inline coalib/output/ConsoleInteraction.py]: this newline is unneeded

**Comment by Adrianzatreanu** [comment]
needs rebase also, and seems tests failing

**Comment by Adrianzatreanu** [comment]
fix tests!!!!! :D

**Comment by Nosferatul** [comment]
I don't know how!!! That's the problem, I tried in several ways

File: coalib/output/ConsoleInteraction.py [hunk 10]
```diff
@@ -54,7 +54,7 @@ def highlight_text(text, lexer=TextLexer(), style=None):
 
 
 STR_GET_VAL_FOR_SETTING = ("Please enter a value for the setting \"{}\" ({}) "
-                           "needed by {}: ")
+                           "needed by {} ")
```
Comment (adtac) [inline coalib/output/ConsoleInteraction.py]: irrelevant change?

File: coalib/output/ConsoleInteraction.py [hunk 11]
```diff
@@ -479,16 +479,20 @@ def require_setting(setting_name, arr):
...
     """
     needed = join_names(arr[1:])
+    section_name = "from section " + "\"" + section.name_str + "\"" + ": "
 
     # Don't use input, it can't deal with escapes!
+
     print(colored(STR_GET_VAL_FOR_SETTING.format(setting_name, arr[0], needed),
-                  REQUIRED_SETTINGS_COLOR))
+                  REQUIRED_SETTINGS_COLOR) + colored(section_name,
+                                                     REQUIRED_SETTINGS_COLOR))
```
Comment (adtac) [inline coalib/output/ConsoleInteraction.py]: Instead change `STR_GET_VAL_FOR_SETTING` to `"Please enter a value for the setting \"{}\" ({}) needed by {} from section \"{}\":"`. And here, use:

```
print(colored(STR_GET_VAL_FOR_SETTING.format(setting_name, arr[0], needed, section.name_str),
                  REQUIRED_SETTINGS_COLOR))
```

Much much cleaner IMO

File: coalib/settings/Section.py [hunk 12]
```diff
@@ -253,3 +253,10 @@ def delete_setting(self, key):
         :param key: The key of the setting to be deleted
         """
         del self.contents[key]
+
+    @property
```
Comment (adtac) [inline coalib/settings/Section.py]: is this really needed? can't we just access `section.name`?

File: tests/coalaJSONTest.py [hunk 13]
```diff
@@ -51,7 +51,7 @@ def test_fail_acquire_settings(self):
             for msg in output["logs"]:
                 if "During execution, we found that some" in msg["message"]:
                     found = True
-            self.assertTrue(found, "Missing settings not logged")
+            self.assertFalse(found, "Missing settings not logged")
```
Comment (adtac) [inline tests/coalaJSONTest.py]: uh, why is this changed?

File: tests/output/ConsoleInteractionTest.py [hunk 14]
```diff
@@ -151,23 +151,27 @@ def tearDown(self):
     def test_require_settings(self):
         self.assertRaises(TypeError, acquire_settings, self.log_printer, 0)
 
+        curr_section = Section("")
```
Comment (adtac) [inline tests/output/ConsoleInteractionTest.py]: can we have a test that verifies that the "from section Default" is present in the output string?

File: coalib/output/ConsoleInteraction.py (hunk 10 above)
Comment (Nosferatul) [inline coalib/output/ConsoleInteraction.py]: @sils told me

Comment (sils) [inline coalib/output/ConsoleInteraction.py]: removing the colon? Why? The colon makes sense. Don't do what I say unless you have reason to believe it's right :P

File: tests/coalaJSONTest.py (hunk 13 above)
Comment (Nosferatul) [inline tests/coalaJSONTest.py]: My bad

File: tests/output/ConsoleInteractionTest.py (hunk 14 above)
Comment (Nosferatul) [inline tests/output/ConsoleInteractionTest.py]: Something like `self.assertIn("from section Default", output)` ?

File: coalib/output/ConsoleInteraction.py (hunk 10 above)
Comment (Nosferatul) [inline coalib/output/ConsoleInteraction.py]: Sorry, my bad, I misunderstand

File: tests/output/ConsoleInteractionTest.py (hunk 14 above)
Comment (Nosferatul) [inline tests/output/ConsoleInteractionTest.py]: Also I have errors in tests and I don't know how to change them to pass it.

File: coalib/output/ConsoleInteraction.py [hunk 15]
```diff
@@ -54,7 +54,7 @@ def highlight_text(text, lexer=TextLexer(), style=None):
 
 
 STR_GET_VAL_FOR_SETTING = ("Please enter a value for the setting \"{}\" ({}) "
-                           "needed by {}: ")
+                           "needed by {} from section \"{}\":")
```
Comment (adtac) [inline coalib/output/ConsoleInteraction.py]: A little pedantic, but there's a space after the colon ;)

File: coalib/output/ConsoleInteraction.py (hunk 9 above)
Comment (adtac) [inline coalib/output/ConsoleInteraction.py]: reminder, this newline isn't needed

File: tests/output/ConsoleInteractionTest.py (hunk 14 above)
Comment (adtac) [inline tests/output/ConsoleInteractionTest.py]: Yep, another assert statement in `test_fail_acquire_settings` will do. No need for a new function :+1:

File: tests/output/dbus/DbusDocumentTest.py [hunk 16]
```diff
@@ -83,5 +83,5 @@ def test_analyze(self):
         uut.SetConfigFile(self.config_path + "2")
         output = uut.Analyze()
         self.assertEqual(output[0], 255)
-        self.assertEqual(output[1][1]["log_level"], "ERROR")
-        self.assertEqual(output[1][1]["message"], Constants.CRASH_MESSAGE)
+        self.assertEqual(output[1][0]["log_level"], "ERROR")
+        self.assertEqual(output[1][0]["message"], Constants.CRASH_MESSAGE)
```
Comment (adtac) [inline tests/output/dbus/DbusDocumentTest.py]: I don't understand this change. What does this do?

File: tests/output/ConsoleInteractionTest.py (hunk 14 above)
Comment (adtac) [inline tests/output/ConsoleInteractionTest.py]: Good find btw :P I didn't know such a function existed

File: coalib/output/ConsoleInteraction.py (hunk 15 above)
Comment (Nosferatul) [inline coalib/output/ConsoleInteraction.py]: True :)

File: tests/output/dbus/DbusDocumentTest.py (hunk 16 above)
Comment (adtac) [inline tests/output/dbus/DbusDocumentTest.py]: Oh alright. But I think it's best if you confirm this with someone else (I don't understand Dbus stuff). Maybe @AbdealiJK (IIRC he worked on Dbus)?

Comment (Nosferatul) [inline tests/output/dbus/DbusDocumentTest.py]: Oh, I see now, when I use `coala-ci -b=SpaceConsistencyBear' it's printed an error.

Comment (AbdealiJK) [inline tests/output/dbus/DbusDocumentTest.py]: That's interesting, I don't see any reason for this to have changed.
Changing it from 1 -> 0 means that the second log message is now the first log message. Was this changed in recent code?
It may be a good idea to do

import pprint from pprint
pprint (output)

And check which error message is not there now

Comment (Nosferatul) [inline tests/output/dbus/DbusDocumentTest.py]: Ok, I'll try it!

File: tests/output/ConsoleInteractionTest.py (hunk 14 above)
Comment (Nosferatul) [inline tests/output/ConsoleInteractionTest.py]: I read through the code and I don't believe it should be in `test_fail_acquire_settings` there is the output for when an acquired setting is bad

File: coalib/output/ConsoleInteraction.py (hunk 9 above)
Comment (Makman2) [inline coalib/output/ConsoleInteraction.py]: blub :3

File: tests/output/InteractionsTest.py [hunk 17]
```diff
@@ -4,15 +4,17 @@
...
 from coalib.output.printers.LogPrinter import LogPrinter
+from coalib.settings.Section import Section
 
 
 class InteractionsTest(unittest.TestCase):
 
     def test_(self):
         log_printer = LogPrinter(NullPrinter())
+        section = Section("")
         self.assertRaises(TypeError, fail_acquire_settings, log_printer, None)
```
Comment (Makman2) [inline tests/output/InteractionsTest.py]: ah you need to change this line too, a `TypeError` is raised here from supplying not enough arguments :)
So you need to add the section argument here, then the coverage raises to 100% again :+1:

**Comment by Makman2** [comment]
looks good, some minor changes left :)

File: coalib/output/ConsoleInteraction.py [hunk 18]
```diff
@@ -54,7 +54,7 @@ def highlight_text(text, lexer=TextLexer(), style=None):
 
 
 STR_GET_VAL_FOR_SETTING = ("Please enter a value for the setting \"{}\" ({}) "
-                           "needed by {}: ")
+                           "needed by {} from section \"{}\": ")
```
Comment (sils) [inline coalib/output/ConsoleInteraction.py]: not `for` instead of `from`?

File: tests/output/ConsoleInteractionTest.py [hunk 19]
```diff
@@ -151,26 +151,31 @@ def tearDown(self):
...
                                               {"setting": ["help text",
                                                            "SomeBear",
                                                            "AnotherBear",
-                                                           "YetAnotherBear"]}),
+                                                           "YetAnotherBear"]},
+                                              curr_section),
                              {"setting": 2})
 
             self.assertEqual(generator.last_input, 2)
+            self.assertIn('', curr_section.name)
```
Comment (Nosferatul) [inline tests/output/ConsoleInteractionTest.py]: I'll be right back at it in 2 hours. Thx for the review @sils @Makman2 !

File: tests/output/InteractionsTest.py (hunk 17 above)
Comment (Nosferatul) [inline tests/output/InteractionsTest.py]: @Makman2 now is 99.95%, something else that I missed? 😄

Comment (Nosferatul) [inline tests/output/InteractionsTest.py]: If I delete the line `514` mentioned above, I'm getting this error `Warning, treated as error:
/home/ubuntu/coala/coalib/output/ConsoleInteraction.py:docstring of coalib.output.ConsoleInteraction.acquire_settings:18: WARNING: Literal block ends without a blank line; unexpected unindent.`

Comment (Nosferatul) [inline tests/output/InteractionsTest.py]: https://circleci.com/gh/coala/coala/8546?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

**Comment by sils** [comment]
In commit message: `Adds a new property in class Section, name_str. This property returns the name of the current section.`

I didn't find this property, could you correct the message or your commit?

**Comment by sils** [comment]
otherwise good IMO! Marking WIP because commit message misinformation.

**Comment by Nosferatul** [comment]
Done @sils

**Comment by sils** [comment]
hey you ignored this twice https://github.com/coala/coala/pull/2799/commits/ded7537d74a96722f78ab2f0161c184e3f0ea55e#r83087277

**Comment by Nosferatul** [comment]
I wrote about it. Rise an error if I delete that line. https://circleci.com/gh/coala/coala/8546?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

**Comment by Nosferatul** [comment]
I'll try it and also rebase because of the out-of-date

File: coalib/output/ConsoleInteraction.py (hunk 9 above)
Comment (sils) [inline coalib/output/ConsoleInteraction.py]: maybe the newline needs to be above the param then?

File: coalib/output/ConsoleInteraction.py [hunk 20]
```diff
@@ -507,6 +510,8 @@ def acquire_settings(log_printer, settings_names_dict):
         {"UseTabs": ["describes whether tabs should be used instead of spaces",
                      "SpaceConsistencyBear",
                      "SomeOtherBear"]}
```
Comment (sils) [inline coalib/output/ConsoleInteraction.py]: nonononono, two newlines here, no newline before the return annotation

**Comment by sils** [comment]
ack https://github.com/coala/coala/pull/2799/commits/46e86feed47f48cb08851b7f51bff99d7fc4319b

**Comment by sils** [comment]
let's let codecov pass (hopefully)

**Comment by sils** [comment]
rebased it for you in https://github.com/coala/coala/pull/2874
